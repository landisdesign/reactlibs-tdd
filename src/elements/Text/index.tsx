import React from 'react';
import styles from './Text.module.scss';

const textTypes = ['story'] as const;

/**
 * Defines properties for Text component
 * @property type If provided, identifies the styling for this component.
 * Currently only `story` is an option.
 * @property html If provided, overrides the children of this component with any
 * content. **This will render any text as HTML, including unsafe scripts.**
 */
export interface TextProps {
    type?: typeof textTypes[number];
    html?: string;
}

const Text: React.FC<TextProps> = (props) => {
    const {
        type = 'default',
        html,
        children
    } = props;

    const className = styles[type] ?? styles.default;

    return html
        ? <div className={className} dangerouslySetInnerHTML={{ __html: html }} />
        : <div className={className}>{children}</div>
    ;
}

export default Text;
