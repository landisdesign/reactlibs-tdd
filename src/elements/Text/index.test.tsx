import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Text from '.';

test('Children are output', () => {
    const content = <b>Hello</b>;
    const expected = <div className='default'>{content}</div>;
    const wrapper = shallow(<Text>{content}</Text>);

    expect(wrapper.matchesElement(expected)).toEqual(true);
});

test('HTML is output dangerously', () => {
    let tree = renderer.create(<Text html='<b>Hello</b>' />).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(<Text html='<b>Hello</b>'>This is not rendered</Text>).toJSON();
    expect(tree).toMatchSnapshot();
});

test('Styles are applied', () => {
    const content = <b>Hello</b>;
    const expected = <div className='story'>{content}</div>;
    const wrapper = shallow(<Text type='story'>{content}</Text>);

    expect(wrapper.matchesElement(expected)).toEqual(true);
})
