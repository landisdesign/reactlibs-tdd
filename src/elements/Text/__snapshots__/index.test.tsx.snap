// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML is output dangerously 1`] = `
<div
  className="default"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<b>Hello</b>",
    }
  }
/>
`;

exports[`HTML is output dangerously 2`] = `
<div
  className="default"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<b>Hello</b>",
    }
  }
/>
`;
